cores/nRF5/HardwarePWM.cpp

```diff
@@ -53,9 +53,123 @@ HardwarePWM* HwPWMx[] =
 #endif
 };
 
-HardwarePWM::HardwarePWM(NRF_PWM_Type* pwm)
+#if CFG_DEBUG
+bool can_stringify_token(uintptr_t token)
 {
-  _pwm = pwm;
+  uint8_t * t = (uint8_t *)&token;
+  for (size_t i = 0; i < sizeof(uintptr_t); ++i, ++t)
+  {
+    uint8_t x = *t;
+    if ((x < 0x20) || (x > 0x7E)) return false;
+  }
+  return true;
+}
+
+void HardwarePWM::DebugOutput(Stream& logger)
+{
+  const size_t count = arrcount(HwPWMx);
+  logger.printf("HwPWM Debug:");
+  for (size_t i = 0; i < count; i++) {
+    HardwarePWM const * pwm = HwPWMx[i];
+    uintptr_t token = pwm->_owner_token;
+    logger.printf(" || %d:", i);
+    if (can_stringify_token(token)) {
+      uint8_t * t = (uint8_t*)(&token);
+      static_assert(sizeof(uintptr_t) == 4);
+      logger.printf("   \"%c%c%c%c\"", t[0], t[1], t[2], t[3] );
+    } else {
+      static_assert(sizeof(uintptr_t) == 4);
+      logger.printf(" %08x", token);
+    }
+    for (size_t j = 0; j < MAX_CHANNELS; j++) {
+      uint32_t r = pwm->_pwm->PSEL.OUT[j]; // only read it once
+      if ( (r & PWM_PSEL_OUT_CONNECT_Msk) != (PWM_PSEL_OUT_CONNECT_Disconnected << PWM_PSEL_OUT_CONNECT_Pos) ) {
+        logger.printf(" %02x", r & 0x1F);
+      } else {
+        logger.printf(" xx");
+      }
+    }
+  }
+  logger.printf("\n");
+}
+#else
+void HardwarePWM::DebugOutput(Stream& logger) {}
+#endif // CFG_DEBUG
+
+// returns true ONLY when (1) no PWM channel has a pin, and (2) the owner token is nullptr
+bool HardwarePWM::takeOwnership(uintptr_t token)
+{
+  bool notInIsr = !isInISR();
+  if (token == 0) {
+    if (notInIsr) {
+      LOG_LV1("HwPWM", "zero / nullptr is not a valid ownership token (attempted use in takeOwnership)");
+    }
+    return false; // cannot take ownership with nullptr
+  }
+  if (token == this->_owner_token) {
+    if (notInIsr) {
+      LOG_LV1("HwPWM", "failing to acquire ownership because already owned by requesting token (cannot take ownership twice)");
+    }
+  }
+  if (this->_owner_token != 0) {
+    return false;
+  }
+  if (this->usedChannelCount() != 0) {
+    return false;
+  }
+  if (this->enabled()) {
+    return false;
+  }
+  // TODO: warn, but do not fail, if taking ownership with IRQs already enabled
+  // NVIC_GetActive
+
+  // Use C++11 atomic CAS operation
+  uintptr_t newValue = 0U;
+  return this->_owner_token.compare_exchange_strong(newValue, token);
```

Comment:
I still prefer using freeRTOS mutex, since I have no idea how this atomic function is implemented in ARM arch with preempted RTOS with/without ISR.

Reply:
Previously, you were uncomfortable because the CAS operation was using a GNU C++ macro, which you said was "[legacy, mainly targeted Intel CPU](https://github.com/adafruit/Adafruit_nRF52_Arduino/pull/496#discussion_r452718048)".

I definitely looked at rewriting using a mutex.  But, if you try, you will see that it quickly gets much more complicated, especially when you have to consider order of initialization for statics, error paths, etc.

The good news is that C++ has [defined the atomic operation](https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange).  Therefore, since it's ensured to work per the C++ standard, I'm not understanding your concern about ARM or whether preemption is enabled in freeRTOS?

From what I could see, the use of the C++ atomic fully addressed your concerns about "legacy" and "mainly targeted Intel CPU".

Reply:
I am new to these atomic operation, unless somebody could point out how these are implemented on ARM, I am not convinced. I suspected it would enable/disable with cpsie/cpsid which is not as ideal as using RTOS critical API. Anyway, I will do the change myself later on.